packages/query-devtools/src/__tests__/utils.test.ts

```diff
@@ -0,0 +1,139 @@
+        const oldData = { title: 'Hello world', id: 1, createdAt: '2021-01-01' }
+
+        const newData = updateNestedDataByPath(
+          oldData,
+          ['title'],
+          'Brave new world',
+        )
+
+        expect(newData).not.toBe(oldData) // should not be the same reference
+
+        expect(oldData['title']).toEqual('Hello world')
+        expect(oldData['id']).toEqual(1)
+        expect(oldData['createdAt']).toEqual('2021-01-01')
+
+        expect(newData['title']).toEqual('Brave new world')
+        expect(newData['id']).toEqual(1)
+        expect(newData['createdAt']).toEqual('2021-01-01')
+      })
+    })
+
+    describe('nested data', () => {
+      it('should update data correctly', async () => {
+        const oldData = [
+          {
+            title: 'Hello world',
+            id: 1,
+            createdAt: '2021-01-01',
+          },
+          {
+            title: 'Second post',
+            id: 2,
+            createdAt: '2023-04-02',
+          },
+        ]
+
+        const newData = updateNestedDataByPath(
+          oldData,
+          ['1', 'createdAt'],
+          '2023-05-01',
+        )
+
+        expect(newData).not.toBe(oldData) // should not be the same reference
+
+        expect(oldData[0]!['title']).toEqual('Hello world')
+        expect(oldData[0]!['id']).toEqual(1)
+        expect(oldData[0]!['createdAt']).toEqual('2021-01-01')
+        expect(oldData[1]!['title']).toEqual('Second post')
+        expect(oldData[1]!['id']).toEqual(2)
+        expect(oldData[1]!['createdAt']).toEqual('2023-04-02')
+
+        expect(newData[0]['title']).toEqual('Hello world')
+        expect(newData[0]['id']).toEqual(1)
+        expect(newData[0]['createdAt']).toEqual('2021-01-01')
+        expect(newData[1]['title']).toEqual('Second post')
+        expect(newData[1]['id']).toEqual(2)
+        expect(newData[1]['createdAt']).toEqual('2023-05-01')
+      })
+    })
+  })
+
+  describe('deleteNestedDataByPath', () => {
+    it('should delete item from array correctly', async () => {
+      const oldData = ['one', 'two', 'three']
+
+      const newData = deleteNestedDataByPath(oldData, ['1'])
+
+      expect(newData).not.toBe(oldData) // should not be the same reference
+
+      expect(newData[0]).toEqual('one')
+      expect(newData[1]).toEqual('three')
+      expect(newData).toHaveLength(2)
+    })
+  })
+
+  it('should delete nested item from array correctly', async () => {
+    const oldData = [
+      {
+        title: 'Pumpkin pie',
+        id: 1,
+        ingredients: ['pumpkin', 'pie crust', 'sugar', 'spices'],
+      },
+      {
+        title: 'Spaghetti bolonese',
+        id: 2,
+        ingredients: ['spaghetti', 'tomato sauce', 'minced meat'],
+      },
+    ]
+    const newData = deleteNestedDataByPath(oldData, ['1', 'ingredients', '0'])
+
+    expect(newData).not.toBe(oldData) // should not be the same reference
+
+    expect(newData).toMatchInlineSnapshot(`
```

Comment:
Do you like the inline snapshot for this kind of test? IMO it makes it super clear what happens and makes additional tests obsolute (e.g. check all array items and properties manually). If you like it I'd use it in other tests too